### ISSUE TITLE
Problems with new beacons

### ISSUE THREAD

**Issue by S-S-X** [pandorabox-io/pandorabox.io#448]
**Here come copy paste from chat**

Related #430 New active Beacon area effects

- [ ] should probably fix range to 20 for upgraded beacons
- Solution?

- [x] chat spamming "Not enough room to place beacon pointing in -X direction!" :p
is there some specific reason why I cannot place beacon without beam..?
I have chat full of beacon message spam because beacons cant be placed with deployer
- Solution: No action needed, deployer actually works exactly as it should. Just different with old beacon because beam always pointed upwards.

- [x] oh god, now I get formcpec spam without touching anything
I got formspec every second while just walking around. But apparently it can be stopped by opening chat.
- Solution: Caused by some (owned) device hitting beacon and triggering formspec. Most probably deployer or digtron. Probably no action needed, very rare thing.

- [x] beam hitbox seems to be a bit too small, that one was not fixed?
I dont have problems seeing hitbox but hitbox is now too small while with old beacons it was too large.
- Solution: Increased size of hitbox (selection box).

- [x] was there way to place beacon to wall while keeping beam pointing upwards?
yes, currently placing with replacer places upwards
deactivate turn off beam and effect
- Solution: Beacons (including beam) can now be deactivated, afterwards beacon can rotated with a screwdriver to change direction.

- [x] should be able to place just beacon box node without space for beam
allow placing beacons when beam cant fit, now probably need to place
using jumpdrive?
- Solution: Beacons can now be placed anywhere, but can only activate if there is enough room.

- [x] digtron got really confused about beacons, maybe it had old or something but seems it
attempted building violet beacon while it only had green in storage
but did not crash at least, just skipped building
- Solution: No solution known, but most likely caused by the corrective name change from purple to violet. 

- [x] does range 50 mean 25+ and 25-?
range 50 should be same as radius 50
so +/- 50 from beacon node
just not same radius that other mods define. that would be good to fix so that it follows
more common definition
range is actually value - half node from outer edge of beacon box. basically range from center of beacon.
- Solution: Changed 'Range' to 'Effect Radius' in formspec.

- [ ] hmm, some beams are missing here for some reason :/
beams from box, then nothing then beams again
why do some disapear anyway when placed by box
- Solution: Most likely something wrong in the cleanup LBM, will need to investigate further.

- [x] now when I think about it this beacon update with colors / selectable effects causes
one more problem, you cannot trust any of beacons you see to have fly effect and also cant tell effect range from distance.
- Solution: No action needed, see comment below. (https://github.com/pandorabox-io/pandorabox.io/issues/448#issuecomment-588449128)

- [ ] In fancy vendor, spinning textured model [of beam nodes] only draws other side.
Also could look a lot better to have extruded model in vendor like most other items.
- Solution: Possibly something to do with the beam node models, will need to investigate further.

- [x] Cannot build to beacons with sneak rmb
Should be possible to do this.
- Solution: Allow skipping formspec when sneak building.

Ping @OgelGames you could probably best answer some of these right away and then we see if there's something left to fix.

**Comment by OgelGames** [comment]
> * should probably fix range to 20 for upgraded beacons

Any reason why?

> * chat spamming "Not enough room to place beacon pointing in -X direction!" :p
> *  is there some specific reason why I cannot place beacon without beam..?
>  * I have chat full of beacon message spam because beacons cant be placed with deployer

Those should all be fixed now: https://github.com/OgelGames/beacon/commit/a3cc21efc87d5f7ed131fd6ef66f5c2eb5fae88c

> * oh god, now I get formcpec spam without touching anything
> *  I got formspec every second while just walking around. But apparently it can be stopped by opening chat.

Don't know why that is happening...

> * beam hitbox seems to be a bit too small, that one was not fixed?
> *  I dont have problems seeing hitbox but hitbox is now too small while with old beacons it was too large.

I ~should~ will make it bigger...

> * was there way to place beacon to wall while keeping beam pointing upwards?
>  * yes, currently placing with replacer places upwards
>  * deactivate turn off beam and effect

That should also be fixed by the above commit, which makes it possible to rotation the beacon with a screwdriver before activating the beacon.

> * should be able to place just beacon box node without space for beam
> *  allow placing beacons when beam cant fit, now probably need to place using jumpdrive?

Again, should be fixed by the above commit.

> * digtron got really confused about beacons, maybe it had old or something but seems it attempted building violet beacon while it only had green in storage but did not crash at least, just skipped building

Probably caused by the corrective name change from purple to violet.

> * does range 50 mean 25+ and 25-?
>   range 50 should be same as radius 50
>   so +/- 50 from beacon node
>   just not same radius that other mods define. that would be good to fix so that it follows
>   more common definition
>   range is actually value - half node from outer edge of beacon box. basically range from center of beacon.

Yeah, I probably should change that...

> * hmm, some beams are missing here for some reason :/
>   beams from box, then nothing then beams again
>   why do some disapear anyway when placed by box

I'm not sure if I understand this one... 

> * now when I think about it this beacon update with colors / selectable effects causes
>   one more problem, you cannot trust any of beacons you see to have fly effect and also cant tell effect range from distance.

That's an unfortunate side effect, which I don't know of a solution for...

**Comment by SwissalpS** [comment]
> > * hmm, some beams are missing here for some reason :/
> >   beams from box, then nothing then beams again
> >   why do some disapear anyway when placed by box
> 
> I'm not sure if I understand this one...

that was from me: place beacon against wall / ceiling (have not noticed with upward pointing beacons)
After some time there is still the box with base-beam + beam * n but then for some nodes just air then n * beam nodes again.
When moving around sometimes more beams 'evaporate'.

**Comment by SwissalpS** [comment]
> * beam hitbox seems to be a bit too small

size of the beam would make sense.

**Comment by OgelGames** [comment]
> > > * hmm, some beams are missing here for some reason :/
> > >   beams from box, then nothing then beams again
> > >   why do some disapear anyway when placed by box
> > 
> > 
> > I'm not sure if I understand this one...
> 
> that was from me: place beacon against wall / ceiling (have not noticed with upward pointing beacons)
> After some time there is still the box with base-beam + beam * n but then for some nodes just air then n * beam nodes again.
> When moving around sometimes more beams 'evaporate'.

That would be something wrong with the cleanup lbm, but that might be fixed now.

> * beam hitbox seems to be a bit too small, that one was not fixed?
>   I dont have problems seeing hitbox but hitbox is now too small while with old beacons it was too large.

Fixed: https://github.com/OgelGames/beacon/commit/a2d4d2d42b0b5e5213b146def88853b8282f7019

**Comment by S-S-X** [comment]
Added one more item to above list: In fancy vendor, spinning textured model only draws other side.

Of course for things that are issues that must be fixed in code it is probably better to open separate issue and link here, ^ that list was mostly to collect all stuff together for easy evaluation and easily dropping things that do not require any real actions.

**Comment by S-S-X** [comment]
>> should probably fix range to 20 for upgraded beacons
> Any reason why?

Because original green beacon range was 20m, so it would not break fly in areas where beacons have already been installed.

**Comment by S-S-X** [comment]
>> I'm not sure if I understand this one...
>
> now when I think about it this beacon update with colors / selectable effects causes
one more problem, you cannot trust any of beacons you see to have fly effect and also cant tell effect range from distance.

Before  change one could see range of beacon, now you cant see range for next beacon before you either fall or reach it.

**Comment by SwissalpS** [comment]
> so any ideas anyone?

no real solution there except try it on and get used to it. Up until now there were many places where you couldn't see beacons at all and glider ensured that it mostly worked out fine.

With many beacons being close together any HUD-radar solution would just crowd the screen to uselessness. /beacons_effects seems to cover it sufficiently. Maybe add coordinates there in the extended info option.

**Comment by SwissalpS** [comment]
I've just tested 6b1486f and these are my opinions:
- beams should start on place if there is space for them -> keep past UX for least friction.
- at least green beacons should also activate fly on placement to keep the grappling hook effect they had
+ replacer seems to work now with correct orientation (we may need to blacklist beacons though [replacer])
- sonic screwdriver has no effect on box
+ screwdriver manipulates param2
- but param2 and direction don't feel tied yet: place box, rotate with screwdriver, re-activate beam -> beam points in same direction no matter how many times you repeat screwdriver
- constructor seems to only be able to take/place vertical beams, not a deal-breaker rn.
+ selection box size is fine now
+ radius is standard now for non green beacons
- 'range' should be 'radius' in formspec
+ green beacons set to 10, still grant fly to 20 which is fine with me as long as they also instantly activate as UX has been.
+ can now sneak+place onto :D

Good job, I like the remake a lot :)

**Comment by OgelGames** [comment]
> Because original green beacon range was 20m, so it would not break fly in areas where beacons have already been installed.

I'm not sure what the issue is here, all the old green beacons will have been converted to a range of 30, and new beacons can be upgraded if 20 is needed.

> Up until now there were many places where you couldn't see beacons at all and glider ensured that it mostly worked out fine.

Very good point :+1: 

> * beams should start on place if there is space for them -> keep past UX for least friction.

I'm fine with that if the beacons are added to the replacer blacklist, because that's the only reason I changed it.

> * at least green beacons should also activate fly on placement to keep the grappling hook effect they had

I could add that, but I'd rather not because then some players might not realize that the beacons have changed.

> * replacer seems to work now with correct orientation (we may need to blacklist beacons though [replacer])

Is it possible to only blacklist it for the multi-node modes, and still be able to use it in single node mode?

> * sonic screwdriver has no effect on box

Interesting...

> * screwdriver manipulates param2
> * but param2 and direction don't feel tied yet: place box, rotate with screwdriver, re-activate beam -> beam points in same direction no matter how many times you repeat screwdriver

It was working fine for me, though the beacon does lack a rotation indicator (all sides are the same).

> * constructor seems to only be able to take/place vertical beams, not a deal-breaker rn.

That's the constructor's problem, not beacons, and the beam isn't really supposed to be diggable anyway :smile: 

> * 'range' should be 'radius' in formspec

I will be fixing that :wink: 

> * green beacons set to 10, still grant fly to 20

That's because the code for the old fly effect is still active in the `pandorabox_custom` mod.

> Edit: for UX sake, it may be good to require a beacon base to be present for effect as it has been.

I will probably change it to that, it makes the beam more important.

**Comment by SwissalpS** [comment]
> > * at least green beacons should also activate fly on placement to keep the grappling hook effect they had
> 
> I could add that, but I'd rather not because then some players might not realize that the beacons have changed.

they should not notice the change by the beacon failing them

**Comment by SwissalpS** [comment]
> Is it possible to only blacklist it for the multi-node modes, and still be able to use it in single node mode?

Yes by maintaining a second list and offering registration methods.

**Comment by BuckarooBanzay** [comment]
> That's because the code for the old fly effect is still active in the pandorabox_custom mod.

Oh, i completely forgot that... i'll remove that part if thats ok...?

**Comment by OgelGames** [comment]
> i'll remove that part if thats ok...?

Yes, should be fine to remove it...

**Comment by S-S-X** [comment]
>> green beacons set to 10, still grant fly to 20
>
> That's because the code for the old fly effect is still active in the pandorabox_custom mod.

That I noticed too but was sure that customization will be removed very soon because it was planned during mod building.

That's why I added this here because it might make some current building sites harder to fix and falling is not really nice way to introduce upgrades.

>> at least green beacons should also activate fly on placement to keep the grappling hook effect they had
>
> I could add that, but I'd rather not because then some players might not realize that the beacons have changed

I bet there's enough things already that will reveal upgrades like placing beacon to wall instead of ground is actually very big and noticeable change, maybe not immediately but will be noticed by players who regularly use beacons.

Also players talk a lot about beacons in chat because they did notice updates and because it is big change.

**Comment by OgelGames** [comment]
I've made a few more changes to fix most of the issues. (also see original post above)

> * beams should start on place if there is space for them -> keep past UX for least friction.

It has been changed to the original action of activating after placement if there is enough room, and a check has been added to prevent activation if placed by a `replacer`.

> * at least green beacons should also activate fly on placement to keep the grappling hook effect they had

I have added a setting (`beacon_default_effect`) to allow changing the effect set on placement, which will fix this if set to `beacon_default_effect = fly`.
